### PR TITLE
feat: add rating controller

### PR DESCRIPTION
**Pull request by 0xff-dev**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it
feat: add rating controller.

#### Which issue(s) this PR fixes

Fix: https://github.com/kubebb/core/issues/210

#### Special notes for your reviewer
**example 1**

one pipeline with two tasks.
```yaml
apiVersion: core.kubebb.k8s.com.cn/v1alpha1
kind: Rating
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"core.kubebb.k8s.com.cn/v1alpha1","kind":"Rating","metadata":{"annotations":{},"name":"rating","namespace":"kubebb-system"},"spec":{"componentName":"kubebb.kubebb-core","pipelineParams":[{"params":[{"name":"URL","value":{"arrayVal":[],"objectVal":{},"stringVal":"https://github.com/kubebb/components/releases/download/kubebb-core-v0.1.10/kubebb-core-v0.1.10.tgz","type":"string"}},{"name":"COMPONENT_NAME","value":{"stringVal":"kubebb-core","type":"string"}},{"name":"VERSION","value":{"stringVal":"v0.1.10","type":"string"}},{"name":"REPOSITORY_NAME","value":{"stringVal":"kubebb","type":"string"}}],"pipelineName":"kubebb"}]}}
  creationTimestamp: "2023-08-23T05:34:18Z"
  generation: 2
  labels:
    rating.component: kubebb.kubebb-core
    rating.repository: kubebb
  name: rating
  namespace: kubebb-system
  resourceVersion: "5531804"
  uid: e987a4ec-3713-43b8-a45d-79ed7feffa34
spec:
  componentName: kubebb.kubebb-core
  pipelineParams:
  - params:
    - name: URL
      value:
        stringVal: https://github.com/kubebb/components/releases/download/kubebb-core-v0.1.10/kubebb-core-v0.1.10.tgz
        type: string
    - name: COMPONENT_NAME
      value:
        stringVal: kubebb-core
        type: string
    - name: VERSION
      value:
        stringVal: v0.1.10
        type: string
    - name: REPOSITORY_NAME
      value:
        stringVal: kubebb
        type: string
    pipelineName: kubebb
status:
  pipelineRuns:
    rating.kubebb:
      actualWeight: 2
      conditions:
      - lastTransitionTime: "2023-08-23T05:39:53Z"
        message: 'Tasks Completed: 2 (Failed: 0, Cancelled 0), Skipped: 0'
        reason: Succeeded
        status: "True"
        type: Succeeded
      expectWeight: 2
      pipelineName: kubebb
      tasks:
      - conditions:
        - lastTransitionTime: "2023-08-23T05:39:53Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        name: kubebb-rback
        taskRunName: rating.kubebb-kubebb-rback
      - conditions:
        - lastTransitionTime: "2023-08-23T05:39:53Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        name: kubebb-helm-lint
        taskRunName: rating.kubebb-kubebb-helm-lint
```

**example 2**
Each of the two pipelines has a task.

```yaml
apiVersion: core.kubebb.k8s.com.cn/v1alpha1
kind: Rating
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"core.kubebb.k8s.com.cn/v1alpha1","kind":"Rating","metadata":{"annotations":{},"name":"two-pipelines","namespace":"kubebb-system"},"spec":{"componentName":"kubebb.kubebb-core","pipelineParams":[{"params":[{"name":"URL","value":{"arrayVal":[],"objectVal":{},"stringVal":"https://github.com/kubebb/components/releases/download/kubebb-core-v0.1.10/kubebb-core-v0.1.10.tgz","type":"string"}},{"name":"COMPONENT_NAME","value":{"stringVal":"kubebb-core","type":"string"}},{"name":"VERSION","value":{"stringVal":"v0.1.10","type":"string"}},{"name":"REPOSITORY_NAME","value":{"stringVal":"kubebb","type":"string"}}],"pipelineName":"only-helm-lint"},{"params":[{"name":"URL","value":{"arrayVal":[],"objectVal":{},"stringVal":"https://github.com/kubebb/components/releases/download/kubebb-core-v0.1.10/kubebb-core-v0.1.10.tgz","type":"string"}},{"name":"COMPONENT_NAME","value":{"stringVal":"kubebb-core","type":"string"}},{"name":"VERSION","value":{"stringVal":"v0.1.10","type":"string"}},{"name":"REPOSITORY_NAME","value":{"stringVal":"kubebb","type":"string"}}],"pipelineName":"only-rback"}]}}
  creationTimestamp: "2023-08-23T05:41:06Z"
  generation: 2
  labels:
    rating.component: kubebb.kubebb-core
    rating.repository: kubebb
  name: two-pipelines
  namespace: kubebb-system
  resourceVersion: "5532807"
  uid: 9bc38934-45e2-4a1a-b3c3-da59c05f7acf
spec:
  componentName: kubebb.kubebb-core
  pipelineParams:
  - params:
    - name: URL
      value:
        stringVal: https://github.com/kubebb/components/releases/download/kubebb-core-v0.1.10/kubebb-core-v0.1.10.tgz
        type: string
    - name: COMPONENT_NAME
      value:
        stringVal: kubebb-core
        type: string
    - name: VERSION
      value:
        stringVal: v0.1.10
        type: string
    - name: REPOSITORY_NAME
      value:
        stringVal: kubebb
        type: string
    pipelineName: only-helm-lint
  - params:
    - name: URL
      value:
        stringVal: https://github.com/kubebb/components/releases/download/kubebb-core-v0.1.10/kubebb-core-v0.1.10.tgz
        type: string
    - name: COMPONENT_NAME
      value:
        stringVal: kubebb-core
        type: string
    - name: VERSION
      value:
        stringVal: v0.1.10
        type: string
    - name: REPOSITORY_NAME
      value:
        stringVal: kubebb
        type: string
    pipelineName: only-rback
status:
  pipelineRuns:
    two-pipelines.only-helm-lint:
      actualWeight: 1
      conditions:
      - lastTransitionTime: "2023-08-23T05:41:16Z"
        message: 'Tasks Completed: 1 (Failed: 0, Cancelled 0), Skipped: 0'
        reason: Succeeded
        status: "True"
        type: Succeeded
      expectWeight: 1
      pipelineName: only-helm-lint
      tasks:
      - conditions:
        - lastTransitionTime: "2023-08-23T05:41:16Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        name: helm-lint
        taskRunName: two-pipelines.only-helm-lint-helm-lint
    two-pipelines.only-rback:
      actualWeight: 1
      conditions:
      - lastTransitionTime: "2023-08-23T05:43:26Z"
        message: 'Tasks Completed: 1 (Failed: 0, Cancelled 0), Skipped: 0'
        reason: Succeeded
        status: "True"
        type: Succeeded
      expectWeight: 1
      pipelineName: only-rback
      tasks:
      - conditions:
        - lastTransitionTime: "2023-08-23T05:43:26Z"
          message: All Steps have completed executing
          reason: Succeeded
          status: "True"
          type: Succeeded
        name: rback
        taskRunName: two-pipelines.only-rback-rback
```

### CODE REVIEW

**Comment by 0xff-dev** [comment]
Does example-test need to be submitted together? In fact, the example-test has not been written yet.😅

File: api/v1alpha1/rating.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+// Before creating pipelinerun, we shoulde delete all the existing pipelineruns.
+func DeletePieline(ctx context.Context, c client.Client, instance *Rating) error {
+	for _, pipelineDef := range instance.Spec.PipelineParams {
+		name := pipelineDef.PipelineName
+		pipelineRun := v1beta1.PipelineRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: instance.Namespace,
+				Name:      fmt.Sprintf("%s.%s", instance.GetName(), name),
```
Comment (Abirdcfly) [inline api/v1alpha1/rating.go:75]: IMO `fmt.Sprintf("%s.%s", instance.GetName(), name)` can be one func.

File: api/v1alpha1/rating.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+	inCreator = func(index int) error {
+		if index >= len(instance.Spec.PipelineParams) {
+			return nil
+		}
+		pipelineDef := instance.Spec.PipelineParams[index]
+		ppr := v1beta1.PipelineRun{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: instance.GetNamespace(),
+				Name:      fmt.Sprintf("%s.%s", instance.GetName(), pipelineDef.PipelineName),
+				OwnerReferences: []metav1.OwnerReference{
```
Comment (Abirdcfly) [inline api/v1alpha1/rating.go:106]: use `controllerutil.SetOwnerReference`?

**Review by Abirdcfly — APPROVED** [review]

**Comment by Abirdcfly** [comment]
rating config/sample and example-test can add by other pr

**Review by Abirdcfly — APPROVED** [review]

File: api/v1alpha1/rating.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+const (
+	RatingComponentLabel  = "rating.component"
+	RatingRepositoryLabel = "rating.repository"
+
+	PipelineRun2RatingLabel     = "rating.pipelinerun"
+	PipelineRun2ComponentLabel  = "rating.pipelinerun.component"
+	PipelineRun2RepositoryLabel = "rating.pipelinerun.repository"
+)
+
+func PipelineRunName(ratingtName, pipelineName string) string {
```
Comment (bjwswang) [inline api/v1alpha1/rating.go:48]: what is `ratingtName`?  a typo?

**Comment by bjwswang** [comment]
> Does example-test need to be submitted together? In fact, the example-test has not been written yet.😅

it's ok to put it in another pr.BUT...make sure this works for the very common case

**Comment by bjwswang** [comment]
> 

Seems the `taskRunName` is wrong

**Review by Abirdcfly — APPROVED** [review]

**Review by bjwswang — APPROVED** [review]

File: api/v1alpha1/rating.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+			p.Value.ObjectVal = param.Value.ObjectVal
+		}
+		result[idx] = p
+	}
+	return result
+}
+
+// Before creating pipelinerun, we shoulde delete all the existing pipelineruns.
+func DeletePieline(ctx context.Context, c client.Client, instance *Rating) error {
```
[the remark this answers is not in the document]
Comment (0xff-dev) [reply api/v1alpha1/rating.go:122]: Yes, `DeletePipelineRun`.

[the remark this answers is not in the document]
Comment (bjwswang) [reply api/v1alpha1/rating.go:122]: pls fix this in your next pr

**Pull request merged**